### PR TITLE
Synthetic Graph Data Generator

### PR DESCRIPTION
**Pull request by MisterPNP**
- add synthetic graph data generator
- add example notebook+data
- add tests for graph data generator

```
---------- coverage: platform darwin, python 3.8.13-final-0 ----------
Name                                     Stmts   Miss  Cover   Missing
----------------------------------------------------------------------
synthetic_data/__init__.py                   0      0   100%
synthetic_data/graph_synthetic_data.py      85     12    86%   26, 29-31, 33, 78, 80, 82, 84, 86, 99, 115
synthetic_data/marginal_dist.py             49      0   100%
synthetic_data/parser.py                    40     17    58%   39, 43-48, 52, 58, 68-82
synthetic_data/synthetic_data.py           126     32    75%   73, 303-305, 354-414
synthetic_data/utils.py                     16      3    81%   16-20
----------------------------------------------------------------------
TOTAL                                      316     64    80%
```

### CODE REVIEW

File: synthetic_data/graph_synthetic_data.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
+'''Generate a synthetic graph using a profile'''
+from cmath import nan
+import math
+
+import dataprofiler as dp
+import networkx as nx
+import scipy.stats as st
+import numpy as np
+
+class GraphDataGenerator():
```
Comment (taylorfturner) [inline synthetic_data/graph_synthetic_data.py:10]:
```suggestion
class GraphDataGenerator(object):
```

File: synthetic_data/graph_synthetic_data.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    '''
+    Synthesize graph data from a GraphProfiler or graph profile.
```
Comment (taylorfturner) [inline synthetic_data/graph_synthetic_data.py:12]:
```suggestion
    Synthesize graph data from a graph profile object
```

File: synthetic_data/graph_synthetic_data.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+    params:
+    graph_profile
+
+    params type:
+    graph_profile: GraphProfiler object/graph_profile
```
Comment (taylorfturner) [inline synthetic_data/graph_synthetic_data.py:18]: what's the difference between the two?

File: synthetic_data/graph_synthetic_data.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+    return:
+    None
+    '''
+
+    def __init__(self, graph):
```
[the remark this answers is not in the document]
Comment (taylorfturner) [reply synthetic_data/graph_synthetic_data.py:24]: although..l as I read `L26-L34` maybe we shouldn't change

File: tests/test_graph_synthetic.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+import networkx as nx
+import numpy as np
+import scipy.stats as st
+
+from synthetic_data.graph_synthetic_data import GraphDataGenerator
+
+test_root_path = os.path.dirname(os.path.dirname(os.path.realpath(__file__)))
+
+class TestSyntheticGraphGenerator(unittest.TestCase):
+    @classmethod
```
Comment (taylorfturner) [inline tests/test_graph_synthetic.py:18]: add new line for break

File: tests/test_graph_synthetic.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        ax1.hist(list(data), bins=100)
+        pts = np.linspace(-3, 4)
+        ax2 = ax1.twinx()
+        ax2.set_ylim(0, 1)
+        ax2.plot(pts, distribution_continuous_test.pdf(pts), color='red')
+        plt.show()
+        self.assertTrue(True)
+
+if __name__ == "__main__":
+    unittest.main()
```
Comment (taylorfturner) [inline tests/test_graph_synthetic.py:111]: new line EOF

File: tests/test_graph_synthetic.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+
+        # plot
+        fig, ax1 = plt.subplots()
+        ax1.hist(list(data), bins=100)
+        pts = np.linspace(-3, 4)
+        ax2 = ax1.twinx()
+        ax2.set_ylim(0, 1)
+        ax2.plot(pts, distribution_continuous_test.pdf(pts), color='red')
+        plt.show()
+        self.assertTrue(True)
```
Comment (taylorfturner) [inline tests/test_graph_synthetic.py:108]: is this asserting the `plt` renders without error?

File: synthetic_data/graph_synthetic_data.py [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+
+            for continuous_attribute in self._continuous_attributes:
+                if continuous_attribute is not None:
+                    sample = self.sample_continuous(continuous_attribute)
+                    edge_attributes[continuous_attribute] = sample
+                    
+            for categorical_attribute in self._categorical_attributes:
+                if categorical_attribute is not None:
+                    sample = self.sample_categorical(categorical_attribute)
+                    edge_attributes[categorical_attribute] = sample
```
Comment (taylorfturner) [inline synthetic_data/graph_synthetic_data.py:62]: we can condense this into one `for` loop. loop over a list of the variables and then in the nested `for` loop, loop over the contents of the respective variable 
```
for attribute in [continuous_attribute, categorical_attribute]:
> for value in attribute:
...
```

**Review by taylorfturner — CHANGES_REQUESTED** [review]

**Review by taylorfturner — APPROVED** [review]

**Review by tazitoo — APPROVED** [review]

**Pull request merged**